src/backend/kvm/thread.rs

```diff
@@ -117,37 +101,67 @@ impl<P: KeepPersonality> super::super::Thread for Thread<P> {
-
-                // To avoid clashing of rep and req in the union, clone the request
-                let req = unsafe { block.msg.req };
-
-                let ret = match i64::from(req.num) {
-                    SYS_ENARX_BALLOON_MEMORY => {
-                        let rep = self.balloon(&req).into();
-                        block.msg.rep = rep;
-                        Ok(Command::Continue)
+                let block: Block = unsafe {
+                    &mut *block_virt.as_mut_ptr::<[usize; BLOCK_SIZE / size_of::<usize>()]>()
+                }
+                .as_mut_slice()
+                .into();
+
+                sallyport::host::execute(block.into_iter().filter_map(|item| {
+                    if cfg!(feature = "dbg")
+                        && !matches!(&item, Item::Syscall(syscall, ..) if syscall.num == 1)
+                        && !matches!(&item, Item::Gdbcall(..))
+                    {
+                        dbg!(&item);
                     }
-
-                    SYS_ENARX_MEM_INFO => {
-                        block.msg.rep = self.meminfo(block).into();
-                        Ok(Command::Continue)
+                    match item {
+                        #[cfg(feature = "gdb")]
+                        Item::Gdbcall(gdbcall, data) => {
+                            unsafe {
+                                execute_gdb(
+                                    gdbcall,
+                                    data,
+                                    &mut self.gdb_fd,
+                                    _gdblisten.as_ref().unwrap(),
+                                )
+                                .unwrap();
+                            }
+                            None
+                        }
+
+                        Item::Enarxcall(enarxcall, data) => match enarxcall {
+                            item::Enarxcall { num, ret, .. }
+                                if *num == item::enarxcall::Number::MemInfo =>
+                            {
+                                let res = self.meminfo();
+                                *ret = match res {
+                                    Ok(n) => n as usize,
+                                    Err(e) => -e as usize,
+                                };
+                                None
+                            }
+
+                            item::Enarxcall {
+                                num,
+                                argv: [log2, npgs, addr, ..],
+                                ret,
+                            } if *num == item::enarxcall::Number::BalloonMemory => {
+                                let res = self.balloon(*log2, *npgs, *addr);
+                                *ret = match res {
+                                    Ok(n) => n as usize,
+                                    Err(e) => -e as usize,
+                                };
+                                None
+                            }
```

Comment:
I think it would be a bit cleaner to have `execute_enarx` or `execute_enarxcall`, same as we have `execute_gdb`

Reply:
done